### ISSUE TITLE
Unable to promptly perform On-Duty task

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1372]
**Describe the bug**
- Currently, if a task has a *Any-Hour* scope, it gets done faster than the *On-Duty* and *Off-Duty* task.
- Therefore, it takes a long time before a "Budget Essential Resources..." type of task to get done. 

**Expected and unexpected behavior**
- *On-Duty* task should have number 1 priority. Next is *Any-Hour* task. Next is *Off-Duty* task.

**Java classes involved**
- `SettlementTask`

**To Reproduce (Please provide steps)**
- This situation can be observed in Backlog task tab in Monitor Tool.

**Screenshots**
- The tasks with the *Any* scope get picked very fast.
- It seems the tasks with the *On-Duty* scope get picked the slowest.
![image](https://github.com/user-attachments/assets/353f0448-1aa1-46bd-bd99-57533355fe7b)

**Specifications**
- pre-3.8.0

**Additional context**
- Before this [commit](https://github.com/mars-sim/mars-sim/commit/692e13d449fb9fb4308469f23619d7153eb2f217), somehow, tend Crop used to be an *On-Duty* task. It rarely got picked by a settler, not even done by gardenbots. For some reasons, all 3 gardenbots are mostly on PowerSave or Wireless Charging and did not get to tend the crops.
- After the change, tend Crop became a *Any-Hour* task. It gets picked very fast, almost instantly.

**Comment by mokun** [comment]
@bevans2000 

This issues that have been lingering around for a while.

Can you look into it ?

**Comment by bevans2000** [comment]
Yes I can look but the priority based onTaskScope will go against the randomly selection based on Score. Let me have a think

**Comment by mokun** [comment]
@bevans2000 

Also, I don't exactly like the backlog tab almost always show the dig local regolith or ice task. 

Each settlement should have a specific quantity of regolith or ice in mine to collect for each day. 

When that quota is metm, it should stop requesting anyone to go out. 

Can we implement a quota system or some sort for regolith/ice ?

**Comment by bevans2000** [comment]
A quote would make sense but it should be driven by the Demand score.

**Comment by mokun** [comment]
Now, it easily reaches 1000 points and max it out. Then it stays at 1000 all the time. Something's not right.

**Comment by bevans2000** [comment]
Let's not do this in 3.8.x as it needs a clear design to change how Tasks are selected

**Comment by bevans2000** [comment]
Relates to #1112

**Comment by mokun** [comment]
@bevans2000 

In conjunction with the On-duty task, we should streamline the use of assignTask() method.

One goal is develop a standard way of querying if a person is capable of starting a task PRIOR to calling assignTask(). 

Currently, the assignTask() requires the task instance as a parameter. This doesn't make sense. 

A task is created before passing the criteria inside the assignTask().

**Comment by bevans2000** [comment]
The point is a Task should not have found it into the Workers potential task list in the first place. The meta task should only score tasks that a worker can do. Sounds like there is a mismatch between some Task and MetaTasks about the criteria.

**Comment by mokun** [comment]
Do you recall the use of assignTask() in some Mission related class (RoverMission, AbstractVehicleMission, etc.) ? 

That's what I was talking about.

MetaTask was not used in evaluating if loading a vehicle cargo is needed. We went straight to starting a Task instance and calling assignTask() right afterward. That's what I don't like.

**Comment by bevans2000** [comment]
Ok so you mean when we directly assign? Yes I can see that maybe a problem.
That needs to be tackled on an individual basis. Maybe we can define a pattern that aly Task that does not have a Meta Task with a score needs a static method is valid(Worker) ?

**Comment by mokun** [comment]
> Ok so you mean when we directly assign? Yes I can see that maybe a problem.

For instance, in DroneMission, we have 
```java
	/**
	 * Assigns a person the task from unloading the drone.
	 *
	 * @param p
	 * @param drone
	 */
	private boolean assignUnloadingCargo(Person person, Drone drone) {
		boolean result = false;
		
		if (person.getAssociatedSettlement().getBuildingManager().addToGarage(drone)) {
			result = assignTask(person, new UnloadVehicleGarage(person, drone));
		} else if (!EVAOperation.isGettingDark(person) && !person.isSuperUnfit()) {
			result = assignTask(person, new UnloadVehicleEVA(person, drone));
		}

		return result;
	}

```

**Comment by mokun** [comment]
This pattern of creating an instance of a task without going through its metatask is used 25x with assignTask() for Person :
 
![image](https://github.com/user-attachments/assets/216512a7-85d5-4855-a4f9-f4835acdd3b9)

Another example in RoverMission's performDepartingFromSettlementPhase : 

```java
			if (member instanceof Person person) {
				// If person has not aboard the rover, board the rover and be ready to depart.
				if (!getRover().isCrewmember(person)) {

					WalkingSteps walkingSteps = new WalkingSteps(person, adjustedLoc, v);
					boolean canWalk = Walk.canWalkAllSteps(person, walkingSteps);
					
					if (canWalk) {
						boolean canDo = assignTask(person, new Walk(person, walkingSteps));
						if (!canDo) {
							logger.warning(person, 10_000, "Unable to start walking toward " + v + ".");
						}
					}

					else { // this crew member cannot find the walking steps to enter the rover
						logger.warning(member, Msg.getString("RoverMission.log.unableToEnter", //$NON-NLS-1$
								v.getName()));
					}
				}
			}
```

In the case above, last year or so, I've modified it to be at least a bit safer to call canWalkAllSteps() prior to creating the Walk instance first, since we don't have a MetaWalk class and that was an excuse. 

But with other tasks, we could have called its metaTasks, or use the settlement task to recruit potential people. 

May be you can change all that in your new project mission architecture.

**Comment by mokun** [comment]
@bevans2000 

Are we still on track for this ?

**Comment by bevans2000** [comment]
At the moment not looking at it. But as I said this is a dangerous area and we should not be forcing the Task manager to select certain Tasks.
The cause is the priority of Tasks is not balanced

**Comment by bevans2000** [comment]
Repair tasks are an Any Hour task so if we did this hard prioritation repairs would never be fixed

**Comment by mokun** [comment]
> Repair tasks are an Any Hour task so if we did this hard prioritation repairs would never be fixed

I don't recall using assignTask() to repair a building/vehicle.

So in terms of repair, I haven't changed anything.

To deconflict what we really want to achieve in this thread, may be I'll start a new issue for dealing with calling assignTask().

**Comment by mokun** [comment]
@muybien3433 
Can you by any chance see the problem I was originally documenting at the start of this thread ?

@bevans2000 
See if you can note how the On-Duty tasks are picked and get done are not optimized.

**Comment by muybien3433** [comment]
> @muybien3433 Can you by any chance see the problem I was originally documenting at the start of this thread ?

Perhaps we could refactor SettlementMetaTask by changing it from an interface to an abstract class and implement some sort method for task list or even create sorted list from which tasks would be picked. 

Also I think we could get rid of assessRobotSuitability because it's used only once.

**Comment by bevans2000** [comment]
So some meta task instances a factory and settlement meta tasks so that is why it's a interface. Also robots do pickup settlement tasks so that method is needed.
If we do sorting, which I'm against, it should be in the Task manager code. If the scoring is not correct, then same tasks will always be at the top of the list and others will get starved.
The root cause is the scoring is not balanced.

**Comment by mokun** [comment]
> The root cause is the scoring is not balanced.

We could add a 3 flags on top of scoring for tasks.

Flags:
1. Emergency
2. Priority
3. Normal

Emergency tasks will always be chosen before priority and normal.
Some priority tasks will be chosen before normal.

For instance:
Eating will be elevated to priority if hunger is at, say, 750 msols.
Eating will become emergency if hunger is at, say, 1500 msols.
Otherwise, eating task is given a normal flag.

**Comment by bevans2000** [comment]
Since we changed the approach to loading and unloading the assignTask is not as much used.

**Comment by mokun** [comment]
> we changed the approach to loading and unloading the assignTask is not as much used.

Right. We've made some changes already.

Also, now, it won't instantiate a new task unless it passes all the checks that it's possible for a person to take on that task first.

We now have checkReplaceTask and replaceTask now to handle that.

```java
	/**
	 * Checks to see if it's okay to replace a task.
	 * 
	 * @param newTask the task to be executed
	 * @param allowSameTask is it allowed to execute the same task as previous
	 */
	public boolean checkReplaceTask(Task newTask, boolean allowSameTask) {
		
		if (newTask == null) {
			return false;
		}
		
		if (hasActiveTask()) {
			
			String currentDes = currentTask.getDescription();

			if (!currentTask.isInterruptable())
				return false;
			
			if (!allowSameTask) {
				if (newTask.getDescription().equalsIgnoreCase(currentDes))
					return false;	
			
		
				if (newTask.getName().equals(getTaskName())) {
					return false;
				}
			}
		}
		
		// Records current task as last task and replaces it with a new task.
		replaceTask(newTask);
		
		return true;
	}
	
	/**
	 * Records the current task as last task and replaces it with a new task.
	 * 
	 * @param newTask
	 */
	public void replaceTask(Task newTask) {
		if ((newTask == null) || newTask.equals(currentTask)) {
			return;
		}
			
		// Backup the current task as last task
		if (currentTask != null)
			lastTask = currentTask;
		
		// Inform that the current task will be terminated
		if (hasActiveTask()) {
			String des = currentTask.getDescription();
			
			currentTask.endTask();
			
			logger.info(worker, 20_000, "Quit '" + des + "' and replace with the new task of '"
						+ newTask.getName() + "'.");
		}
		
		// Make the new task as the current task
		currentTask = newTask;
		currentScore = null;
		
		// Send out the task event
		worker.fireUnitUpdate(UnitEventType.TASK_EVENT, newTask);
	}

```
